[CDAP-4652] - Feature/ui resources pipelinesettings

ajainarayanan opened the pull request:
- Adds `memoryMb` & `virtualCores` resource settings to batch pipelines.
- Fixes `my-card` directive to have `data-title` attribute instead of `title` to avoid browser triggering tooltip.
- Updates `config-store` & `detail-nonruns-store` based on above modifications.

Bamboo build: http://builds.cask.co/browse/CDAP-DRC4214-1
#### Note:
- Right now UI shows a tooltip before `Resources` settings in batch pipelines if the user is in `Standalone` mode. Yet to confirm it with @ToddGreenstein  or @nitinmotgi . It is to indicate that `Resources` settings don't take any effect in standalone mode.

ajainarayanan committed:
Adds initial version of resources to top panel in hydrator studio view

ajainarayanan commented:
~~Yet to add it to other pipeline types. Hence the on-hold~~

EDIT:
Added resources settings for all pipeline types.

elfenheart commented on cdap-ui/app/directives/my-pipeline-settings/my-batch-pipeline-settings/my-batch-pipeline-settings.html:
```diff
@@ -69,11 +69,40 @@
+          </div>
+          <div class="resource-holder" ng-if="MyBatchPipelineSettingsCtrl.showResources">
+            <div class="resource-group">
+              <label class="control-label"> Virual Cores </label>
```
Typo: Virtual

elfenheart commented on cdap-ui/app/directives/my-pipeline-settings/my-realtime-pipeline-settings/my-realtime-pipeline-settings.html:
```diff
@@ -24,5 +24,26 @@
+    <my-card data-title="Resources" class="resources">
+      <div class="resource-holder">
+        <div class="resource-group">
+          <label class="control-label"> Virual Cores </label>
```
Typo Virtual

elfenheart commented:
![screen shot 2016-07-29 at 11 21 07 am](https://cloud.githubusercontent.com/assets/4398643/17259056/9544b262-557e-11e6-98a2-3537dd3e6e86.png)

The label doesn't look like it's vertically centered

elfenheart commented:
👍  LGTM

ajainarayanan committed:
Adds pipeline-resource factory to render pipeline resources (both driver and executor)

ajainarayanan committed:
Modifies Gulpfile to add my-pipeline-resource to es6 directives

ajainarayanan committed:
Fixes my-pipeline-resource directive to share a template across controller (driver & executor resource components)

ajainarayanan committed:
Fixes my-batch-pipeline-settings to remove collapsable sections for resources + adds width to resource(engine) card in the top panel

ajainarayanan committed:
Fixes my-realtime-pipeline-settings directive to not have resources in the top panel

ajainarayanan committed:
Fixes config-store, config-action & pipeline-detail-nonruns-store to get/set resources from configuration

ajainarayanan committed:
Minor styling fixes for pipeline settings in published pipeline view

ajainarayanan committed:
Fixes services config-store and config-action in hydrator studio to have consistent api

ajainarayanan committed:
Adds resources card in data streams pipeline type's top panel

elfenheart commented on cdap-ui/Gulpfile.js:
```diff
@@ -235,14 +235,15 @@ function getEs6Directives(isNegate) {
     'widget-container',
     'plugin-functions',
-    'my-link-button'
+    'my-link-button',
```
trailing comma?

elfenheart commented on cdap-ui/app/directives/my-pipeline-resource/my-pipeline-driver-resource-ctrl.js:
```diff
@@ -0,0 +1,30 @@
+  'ngInject';
+  $scope.virtualCores = $scope.store.getDriverVirtualCores();
+  $scope.memoryMB = $scope.store.getDriverMemoryMB();
+  $scope.cores = Array.apply(null, {length: 21}).map(Number.call, Number).filter(a => a > 0);
```
`.map(function (value, index) {  return index + 1; }`  might be simpler?

ajainarayanan committed:
Minor fixes in Gulpfile, my-pipeline-driver-resource & my-pipeline-executor-resource controllers

elfenheart commented on cdap-ui/app/directives/my-pipeline-resource/my-pipeline-executor-resource-ctrl.js:
```diff
@@ -0,0 +1,29 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+*/
+function MyPipelineExecutorResourceCtrl($scope) {
```
do these need to be separate directives? it seemed like it's doing the same thing?

ajainarayanan replied:
So its essentially two directives but sharing the same template. The template is same. The controller needs to update different properties in the state.

elfenheart commented on cdap-ui/app/directives/my-pipeline-resource/my-pipeline-resource.html:
```diff
@@ -0,0 +1,32 @@
+</div>
+<div class="resource-group">
+  <label class="control-label"> Memory </label>
+  <input class="form-control"
```
we're not giving any warning if they input invalid input anymore? such as negative number?

ajainarayanan committed:
Fixes number widget to handle empty or negative integers for pipeline resources + minor styling fixes to be consistent with other error notifications.
- Adds default resources (driver + executor) on import/pipeline template use cases in hydrator when the resources are not set in config
- Fixes number widget to handle empty + negative numbers
- Adds validation for pipeline resources (empty & negative integers)
- Fixes error styling in number widget
- Adds the message to indicate error in pipeline resources
- Fixes my-pipeline-resource to have isDisabled attribute in scope

elfenheart commented:
Importing a pipeline without resources does not set the default

elfenheart commented:
Existing pipeline that does not have resource show empty boxes

elfenheart commented:
👍  LGTM

ajainarayanan committed:
Updates missing values message + fixes default value in the error tooltip

ajainarayanan commented:
Demoed it to @ToddGreenstein Merging.
